From aed5371dd689a620b1d907cdbab5b3a89fbc3627 Mon Sep 17 00:00:00 2001
From: Huacai Chen <chenhc@lemote.com>
Date: Sat, 23 Jun 2018 14:14:24 +0800
Subject: [PATCH 289/317] AOSCOS: parport: Add support for the WCH384 4S/1P
 multi-IO card

WCH384 4S/1P is a PCI-E card with 1 LPT and 4 DB9 COM ports detected as
Serial controller: Device 1c00:3450 (rev 10) (prog-if 05 [16850])

[Mingcong Bai: Resolved minor merge conflicts in
drivers/tty/serial/8250/8250_pci.c.]

Signed-off-by: Huacai Chen <chenhc@lemote.com>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/parport/parport_serial.c   | 10 ++++++++++
 drivers/tty/serial/8250/8250_pci.c | 10 ++++++++++
 2 files changed, 20 insertions(+)

diff --git a/drivers/parport/parport_serial.c b/drivers/parport/parport_serial.c
index 3644997a8342..d09d32819609 100644
--- a/drivers/parport/parport_serial.c
+++ b/drivers/parport/parport_serial.c
@@ -60,6 +60,7 @@ enum parport_pc_pci_cards {
 	wch_ch353_2s1p,
 	wch_ch382_0s1p,
 	wch_ch382_2s1p,
+	wch_ch384_4s1p,
 	brainboxes_5s1p,
 	sunix_4008a,
 	sunix_5069a,
@@ -157,6 +158,7 @@ static struct parport_pc_pci cards[] = {
 	/* wch_ch353_2s1p*/             { 1, { { 2, -1}, } },
 	/* wch_ch382_0s1p*/		{ 1, { { 2, -1}, } },
 	/* wch_ch382_2s1p*/             { 1, { { 2, -1}, } },
+	/* wch_ch384_4s1p*/             { 1, { { 2, -1}, } },
 	/* brainboxes_5s1p */           { 1, { { 3, -1 }, } },
 	/* sunix_4008a */		{ 1, { { 1, 2 }, } },
 	/* sunix_5069a */		{ 1, { { 1, 2 }, } },
@@ -270,6 +272,7 @@ static struct pci_device_id parport_serial_pci_tbl[] = {
 	{ 0x4348, 0x7053, 0x4348, 0x3253, 0, 0, wch_ch353_2s1p},
 	{ 0x1c00, 0x3050, 0x1c00, 0x3050, 0, 0, wch_ch382_0s1p},
 	{ 0x1c00, 0x3250, 0x1c00, 0x3250, 0, 0, wch_ch382_2s1p},
+	{ 0x1c00, 0x3450, 0x1c00, 0x3450, 0, 0, wch_ch384_4s1p},
 
 	/* BrainBoxes PX272/PX306 MIO card */
 	{ PCI_VENDOR_ID_INTASHIELD, 0x4100,
@@ -558,6 +561,13 @@ static struct pciserial_board pci_parport_serial_boards[] = {
 		.uart_offset    = 8,
 		.first_offset   = 0xC0,
 	},
+	[wch_ch384_4s1p] = {
+		.flags          = FL_BASE0,
+		.num_ports      = 4,
+		.base_baud      = 115200,
+		.uart_offset    = 8,
+		.first_offset   = 0xC0,
+	},
 	[brainboxes_5s1p] = {
 		.flags		= FL_BASE2,
 		.num_ports	= 5,
diff --git a/drivers/tty/serial/8250/8250_pci.c b/drivers/tty/serial/8250/8250_pci.c
index 3c3f7c926afb..fa3c572f34c6 100644
--- a/drivers/tty/serial/8250/8250_pci.c
+++ b/drivers/tty/serial/8250/8250_pci.c
@@ -79,6 +79,7 @@
 #define PCIE_VENDOR_ID_WCH		0x1c00
 #define PCIE_DEVICE_ID_WCH_CH382_2S1P	0x3250
 #define PCIE_DEVICE_ID_WCH_CH384_4S	0x3470
+#define PCIE_DEVICE_ID_WCH_CH384_4S1P	0x3450
 #define PCIE_DEVICE_ID_WCH_CH384_8S	0x3853
 #define PCIE_DEVICE_ID_WCH_CH382_2S	0x3253
 
@@ -2895,6 +2896,14 @@ static struct pci_serial_quirk pci_serial_quirks[] = {
 		.subdevice      = PCI_ANY_ID,
 		.init           = pci_wch_ch38x_init,
 		.exit		= pci_wch_ch38x_exit,
+		.setup		= pci_wch_ch38x_setup,
+	},
+	/* WCH CH384 4S1P card (16850 clone) */
+	{
+		.vendor         = PCIE_VENDOR_ID_WCH,
+		.device         = PCIE_DEVICE_ID_WCH_CH384_4S1P,
+		.subvendor      = PCI_ANY_ID,
+		.subdevice      = PCI_ANY_ID,
 		.setup          = pci_wch_ch38x_setup,
 	},
 	/*
@@ -3972,6 +3981,7 @@ static const struct pci_device_id blacklist[] = {
 	{ PCI_DEVICE(0x4348, 0x5053), 0, 0, REPORT_CONFIG(PARPORT_SERIAL), },
 	/* WCH CH382 2S1P */
 	{ PCI_DEVICE(0x1c00, 0x3250), 0, 0, REPORT_CONFIG(PARPORT_SERIAL), },
+	{ PCI_DEVICE(0x1c00, 0x3450), }, /* WCH CH384 4S1P */
 
 	/* Intel platforms with MID UART */
 	{ PCI_VDEVICE(INTEL, 0x081b), REPORT_8250_CONFIG(MID), },
-- 
2.48.1

